### ISSUE TITLE
Xaya crashes when restarted

### ISSUE THREAD

**Issue by DarkClaw1** [xaya/xaya#82]
**Describe the issue**
Xaya can only sync from scratch, if xayad/etc is closed and restarted it crashes

**What behavior did you expect?**
To be able to restart the xaya process

**What was the actual behavior (provide screenshots if the issue is GUI-related)?**
**-Linux:**
```
xxx@xxx:~/xaya$ xaya-qt
qt5ct: using qt5ct plugin
xaya-qt: chain.cpp:111: const CBlockIndex* CBlockIndex::GetAncestor(int) const: Assertion `pindexWalk->pprev' failed.
Aborted (core dumped)
xxx@xxx:~/xaya$ xayad
2019-01-23T02:50:38Z Xaya version v1.1 (release build)
2019-01-23T02:50:38Z InitParameterInteraction: parameter interaction: -whitelistforcerelay=1 -> setting -whitelistrelay=1
2019-01-23T02:50:38Z Assuming ancestors of block e685ccaa62025c5c5075cfee80e498589bd4788614dcbe397e12bf2b8e887e47 have valid signatures.
2019-01-23T02:50:38Z Setting nMinimumChainWork=0000000000000000000000000000000000000000000f469d71a33ceda65de424
2019-01-23T02:50:38Z Using the 'shani(1way,2way)' SHA256 implementation
2019-01-23T02:50:38Z Using RdRand as an additional entropy source
2019-01-23T02:50:38Z Default data directory /home/xxx/.xaya
2019-01-23T02:50:38Z Using data directory /home/xxx/.xaya
2019-01-23T02:50:38Z Config file: /home/xxx/.xaya/xaya.conf (not found, skipping)
2019-01-23T02:50:38Z Using at most 125 automatic connections (1024 file descriptors available)
2019-01-23T02:50:38Z Using 16 MiB out of 32/2 requested for signature cache, able to store 524288 elements
2019-01-23T02:50:39Z Using 16 MiB out of 32/2 requested for script execution cache, able to store 524288 elements
2019-01-23T02:50:39Z Using 16 threads for script verification
2019-01-23T02:50:39Z scheduler thread start
2019-01-23T02:50:39Z HTTP: creating work queue of depth 16
2019-01-23T02:50:39Z No rpcpassword set - using random cookie authentication.
2019-01-23T02:50:39Z Generated RPC authentication cookie /home/xxx/.xaya/.cookie
2019-01-23T02:50:39Z HTTP: starting 4 worker threads
2019-01-23T02:50:39Z Using wallet directory /home/xxx/.xaya/wallets
2019-01-23T02:50:39Z init message: Verifying wallet(s)...
2019-01-23T02:50:39Z Using BerkeleyDB version Berkeley DB 4.8.30: (April  9, 2010)
2019-01-23T02:50:39Z Using wallet wallet.dat
2019-01-23T02:50:39Z BerkeleyEnvironment::Open: LogDir=/home/xxx/.xaya/wallets/database ErrorFile=/home/xxx/.xaya/wallets/db.log
2019-01-23T02:50:39Z Cache configuration:
2019-01-23T02:50:39Z * Using 2.0MiB for block index database
2019-01-23T02:50:39Z * Using 8.0MiB for chain state database
2019-01-23T02:50:39Z * Using 440.0MiB for in-memory UTXO set (plus up to 286.1MiB of unused mempool space)
2019-01-23T02:50:39Z init message: Loading block index...
2019-01-23T02:50:39Z Opening LevelDB in /home/xxx/.xaya/blocks/index
2019-01-23T02:50:39Z Opened LevelDB successfully
2019-01-23T02:50:39Z Using obfuscation key for /home/xxx/.xaya/blocks/index: 0000000000000000
xayad: chain.cpp:111: const CBlockIndex* CBlockIndex::GetAncestor(int) const: Assertion `pindexWalk->pprev' failed.
Aborted (core dumped)
```
**-Windows:**
```
2019-01-23T03:23:37Z Xaya version v1.1.0.0-52bc22138-dirty (release build)
2019-01-23T03:23:37Z InitParameterInteraction: parameter interaction: -whitelistforcerelay=1 -> setting -whitelistrelay=1
2019-01-23T03:23:37Z Assuming ancestors of block e685ccaa62025c5c5075cfee80e498589bd4788614dcbe397e12bf2b8e887e47 have valid signatures.
2019-01-23T03:23:37Z Setting nMinimumChainWork=0000000000000000000000000000000000000000000f469d71a33ceda65de424
2019-01-23T03:23:37Z Using the 'sse4(1way),sse41(4way)' SHA256 implementation
2019-01-23T03:23:38Z Default data directory C:\Users\xxx\AppData\Roaming\Xaya
2019-01-23T03:23:38Z Using data directory C:\Users\xxx\AppData\Roaming\Xaya
2019-01-23T03:23:38Z Config file: C:\Users\xxx\AppData\Roaming\Xaya\xaya.conf
2019-01-23T03:23:38Z Using at most 125 automatic connections (2048 file descriptors available)
2019-01-23T03:23:38Z Using 16 MiB out of 32/2 requested for signature cache, able to store 524288 elements
2019-01-23T03:23:38Z Using 16 MiB out of 32/2 requested for script execution cache, able to store 524288 elements
2019-01-23T03:23:38Z Using 12 threads for script verification
2019-01-23T03:23:38Z scheduler thread start
2019-01-23T03:23:38Z HTTP: creating work queue of depth 16
2019-01-23T03:23:38Z No rpcpassword set - using random cookie authentication.
2019-01-23T03:23:38Z Generated RPC authentication cookie C:\Users\xxx\AppData\Roaming\Xaya\.cookie
2019-01-23T03:23:38Z HTTP: starting 4 worker threads
2019-01-23T03:23:38Z Using wallet directory C:\Users\xxx\AppData\Roaming\Xaya\wallets
2019-01-23T03:23:38Z init message: Verifying wallet(s)...
2019-01-23T03:23:38Z Using BerkeleyDB version Berkeley DB 4.8.30: (April  9, 2010)
2019-01-23T03:23:38Z Using wallet wallet.dat
2019-01-23T03:23:38Z BerkeleyEnvironment::Open: LogDir=C:\Users\xxx\AppData\Roaming\Xaya\wallets\vault.dat\database ErrorFile=C:\Users\xxx\AppData\Roaming\Xaya\wallets\vault.dat\db.log
2019-01-23T03:23:38Z Using BerkeleyDB version Berkeley DB 4.8.30: (April  9, 2010)
2019-01-23T03:23:38Z Using wallet wallet.dat
2019-01-23T03:23:38Z BerkeleyEnvironment::Open: LogDir=C:\Users\xxx\AppData\Roaming\Xaya\wallets\game.dat\database ErrorFile=C:\Users\xxx\AppData\Roaming\Xaya\wallets\game.dat\db.log
2019-01-23T03:23:38Z Cache configuration:
2019-01-23T03:23:38Z * Using 2.0MiB for block index database
2019-01-23T03:23:38Z * Using 56.0MiB for transaction index database
2019-01-23T03:23:38Z * Using 8.0MiB for chain state database
2019-01-23T03:23:38Z * Using 384.0MiB for in-memory UTXO set (plus up to 286.1MiB of unused mempool space)
2019-01-23T03:23:38Z init message: Loading block index...
2019-01-23T03:23:38Z Opening LevelDB in C:\Users\xxx\AppData\Roaming\Xaya\blocks\index
2019-01-23T03:23:38Z Opened LevelDB successfully
2019-01-23T03:23:38Z Using obfuscation key for C:\Users\xxx\AppData\Roaming\Xaya\blocks\index: 0000000000000000
2019-01-23T03:23:39Z ThreadRPCServer incorrect password attempt from 127.0.0.1:57395
2019-01-23T03:23:39Z ThreadRPCServer incorrect password attempt from 127.0.0.1:57397
2019-01-23T03:23:39Z ThreadRPCServer incorrect password attempt from 127.0.0.1:57396

```

**How reliably can you reproduce the issue, what are the steps to do so?**
Simply stop xayad/etc and restart it.

**What version of Xaya Core are you using, where did you get it (website, self-compiled, etc)?**
Version 1.1, both on linux (self-compiled) and windows 7 (packaged with xaya-electron 1.1.1 from https://github.com/xaya/xaya_electron/releases)

**What type of machine are you observing the error on (OS/CPU and disk type)?**
**-Machine 1:**
OS: Linux mint 19.1 (kernel 4.20.0) 
CPU: AMD 2990wx
Disk: SSD

**-Machine 2:**
OS: Windows 7 
CPU: Intel i7 3930k
Disk: SSD

**Further Info**
Other users reporting the same here: https://forum.xaya.io/topic/396-xaya-bug/

**Comment by domob1812** [comment]
Thanks for opening the report!  I believe this is caused by non-zero `nNonce` values in recent blocks (as enabled after the fork).  This is all fine from the consensus point of view, but it causes issues with the on-disk block index.

I'm investigating this further and will test a fix shortly.

**Comment by domob1812** [comment]
Indeed.  The reason for this bug is as follows:  The `nNonce` field was removed from `CBlockIndex` in Xaya, since it used to be always zero - but this is no longer the case after the fork.  The result is that `nNonce` is zero (or actually not present and assumed to be zero) for `CBlockIndex` instances that are loaded from disk during startup - and then the block hash is computed for them to insert into `mapBlockIndex`.  But that block hash is wrong, since it is computed assuming `nNonce == 0`.

Consequently, when the *next* block is loaded, its previous block hash `hashPrev` refers to the correct block hash, but there is no entry for it in `mapBlockIndex`.  Thus, the `pprev` will refer to an almost-empty struct, which then causes the `assert` failure.

**Comment by domob1812** [comment]
Proposed fix is in #83.  We will merge that to `master` and `1.1` after appropriate testing.  But everyone feel free to test themselves with the pull request, of course - any feedback is welcome!

**Comment by DarkClaw1** [comment]
```
xxx@xxx:~$ xayad -reindex
2019-01-23T19:42:20Z Xaya version v1.1.99.0-f064cfd2b (release build)
2019-01-23T19:42:20Z InitParameterInteraction: parameter interaction: -whitelistforcerelay=1 -> setting -whitelistrelay=1
2019-01-23T19:42:20Z Assuming ancestors of block e685ccaa62025c5c5075cfee80e498589bd4788614dcbe397e12bf2b8e887e47 have valid signatures.
2019-01-23T19:42:20Z Setting nMinimumChainWork=0000000000000000000000000000000000000000000f469d71a33ceda65de424
2019-01-23T19:42:20Z Using the 'shani(1way,2way)' SHA256 implementation
2019-01-23T19:42:20Z Using RdRand as an additional entropy source
2019-01-23T19:42:20Z Default data directory /home/xxx/.xaya
2019-01-23T19:42:20Z Using data directory /home/xxx/.xaya
2019-01-23T19:42:20Z Config file: /home/xxx/.xaya/xaya.conf (not found, skipping)
2019-01-23T19:42:20Z Using at most 125 automatic connections (1024 file descriptors available)
2019-01-23T19:42:20Z Using 16 MiB out of 32/2 requested for signature cache, able to store 524288 elements
2019-01-23T19:42:20Z Using 16 MiB out of 32/2 requested for script execution cache, able to store 524288 elements
2019-01-23T19:42:20Z Using 16 threads for script verification
2019-01-23T19:42:20Z scheduler thread start
2019-01-23T19:42:20Z HTTP: creating work queue of depth 16
2019-01-23T19:42:20Z No rpcpassword set - using random cookie authentication.
2019-01-23T19:42:20Z Generated RPC authentication cookie /home/xxx/.xaya/.cookie
2019-01-23T19:42:20Z HTTP: starting 4 worker threads
2019-01-23T19:42:20Z Using wallet directory /home/xxx/.xaya/wallets
2019-01-23T19:42:20Z init message: Verifying wallet(s)...
2019-01-23T19:42:20Z Using BerkeleyDB version Berkeley DB 4.8.30: (April  9, 2010)
2019-01-23T19:42:20Z Using wallet wallet.dat
2019-01-23T19:42:20Z BerkeleyEnvironment::Open: LogDir=/home/xxx/.xaya/wallets/database ErrorFile=/home/xxx/.xaya/wallets/db.log
2019-01-23T19:42:20Z Cache configuration:
2019-01-23T19:42:20Z * Using 2.0MiB for block index database
2019-01-23T19:42:20Z * Using 8.0MiB for chain state database
2019-01-23T19:42:20Z * Using 440.0MiB for in-memory UTXO set (plus up to 286.1MiB of unused mempool space)
2019-01-23T19:42:20Z init message: Loading block index...
2019-01-23T19:42:20Z Wiping LevelDB in /home/xxx/.xaya/blocks/index
2019-01-23T19:42:20Z Opening LevelDB in /home/xxx/.xaya/blocks/index
2019-01-23T19:42:20Z Opened LevelDB successfully
2019-01-23T19:42:20Z Using obfuscation key for /home/xxx/.xaya/blocks/index: 0000000000000000
2019-01-23T19:42:20Z Initializing databases...
2019-01-23T19:42:20Z Wiping LevelDB in /home/xxx/.xaya/chainstate
2019-01-23T19:42:20Z Opening LevelDB in /home/xxx/.xaya/chainstate
2019-01-23T19:42:20Z Opened LevelDB successfully
2019-01-23T19:42:20Z Wrote new obfuscate key for /home/xxx/.xaya/chainstate: 54c2671e299a018b
2019-01-23T19:42:20Z Using obfuscation key for /home/xxx/.xaya/chainstate: 54c2671e299a018b
2019-01-23T19:42:20Z  block index              26ms
2019-01-23T19:42:20Z init message: Loading wallet...
2019-01-23T19:42:20Z [default wallet] nFileVersion = 1019900
2019-01-23T19:42:20Z [default wallet] Keys: 2005 plaintext, 0 encrypted, 2005 w/ metadata, 2005 total. Unknown wallet records: 0
2019-01-23T19:42:20Z [default wallet] Wallet completed loading in              26ms
2019-01-23T19:42:20Z [default wallet] setKeyPool.size() = 2000
2019-01-23T19:42:20Z [default wallet] mapWallet.size() = 3
2019-01-23T19:42:20Z [default wallet] mapAddressBook.size() = 4
2019-01-23T19:42:20Z Reindexing block file blk00000.dat...
xayad: validation.cpp:3229: bool ContextualCheckBlock(const CBlock&, CValidationState&, const Consensus::Params&, const CBlockIndex*): Assertion `pindexPrev != nullptr' failed.
Aborted (core dumped)
```

**Comment by xaya** [comment]
working on windows

@DarkClaw1 did you test on linux?

(yes unfortunately you need to sync from scratch)

**Comment by DarkClaw1** [comment]
> @DarkClaw1 did you test on linux?

Yes, I only tested on linux.

**Comment by domob1812** [comment]
The fix from #83 has been merged to `master` and backported to `1.1`.  It is also present in the latest bug-fix release, https://github.com/xaya/xaya/releases/tag/v1.1.1.
